### PR TITLE
feat(datastore): Add selective sync

### PR DESCRIPTION
**Pull request by cshfang**
Currently, DataStore downloads the entire contents of your cloud data source to local device by default when using Amplify Flutter. In comparison, customers of native Amplify Android and iOS SDKs are able to configure DataStore to selectively persist only a subset of data instead.

This PR adds selective sync to Amplify Flutter. #298

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt [hunk 1]
```diff
@@ -118,31 +131,39 @@ class AmplifyDataStorePlugin : FlutterPlugin, MethodCallHandler {
...
+                        "Received invalid request from Dart, modelSchemas and/or modelProviderVersion" +
+                                " are not available. Request: " + request.toString()
+                    )
+                )
             }
             return
         }
 
         val modelSchemas: List<Map<String, Any>> = request["modelSchemas"].safeCastToList()!!
+        val syncExpressions: List<Map<String, Any>> =
```
Comment (cshfang) [inline packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt:152]: ❗ 152-239

File: packages/amplify_datastore/ios/Classes/SwiftAmplifyDataStorePlugin.swift [hunk 2]
```diff
@@ -77,36 +78,63 @@ public class SwiftAmplifyDataStorePlugin: NSObject, FlutterPlugin {
...
         let syncMaxRecords = args["syncMaxRecords"] as? UInt ?? DataStoreConfiguration.defaultSyncMaxRecords
         let syncPageSize = args["syncPageSize"] as? UInt ?? DataStoreConfiguration.defaultSyncPageSize
-
+        let controller = UIApplication.shared.windows.first { $0.isKeyWindow }!.rootViewController as! FlutterViewController
+        let channel = FlutterMethodChannel(name: "com.amazonaws.amplify/datastore", binaryMessenger: controller.binaryMessenger)
+        
         do {
-
+            
+
```
Comment (cshfang) [inline packages/amplify_datastore/ios/Classes/SwiftAmplifyDataStorePlugin.swift:98]: Will double check if it was xcode that did this

File: packages/amplify_datastore/android/src/test/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePluginTest.kt [hunk 3]
```diff
@@ -151,89 +173,177 @@ class AmplifyDataStorePluginTest {
...
+            verify(
+                mockDataStoreConfigurationBuilder,
+                times(1)
+            ).syncInterval(mockSyncInterval.toLong(), TimeUnit.MINUTES)
             verify(mockDataStoreConfigurationBuilder, times(1)).syncMaxRecords(mockSyncMaxRecords)
             verify(mockDataStoreConfigurationBuilder, times(1)).syncPageSize(mockSyncPageSize)
         }
     }
 
+    @Test
```
Comment (cshfang) [inline packages/amplify_datastore/android/src/test/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePluginTest.kt:191]: ❗ 191 - 245

File: packages/amplify_datastore/ios/Classes/SwiftAmplifyDataStorePlugin.swift [hunk 4]
```diff
@@ -77,36 +78,63 @@ public class SwiftAmplifyDataStorePlugin: NSObject, FlutterPlugin {
...
                 try FlutterModelSchema.init(serializedData: $0).convertToNativeModelSchema()
             }
-
+            
             modelSchemas.forEach { (modelSchema) in
                 flutterModelRegistration.addModelSchema(modelName: modelSchema.name, modelSchema: modelSchema)
             }
-
+            
+            let syncExpressions: [DataStoreSyncExpression] = try syncExpressionList.map { syncExpression in
```
Comment (HuiSF) [inline packages/amplify_datastore/ios/Classes/SwiftAmplifyDataStorePlugin.swift:107]: This `onConfigureDataStore` became quite long now, would that be better to extract the synchronous query predicate resolving logic to a dedicated function? Same for the Android implementation.

File: packages/amplify_datastore/test/amplify_datastore_test.dart [hunk 5]
```diff
@@ -40,21 +48,31 @@ void main() {
...
-        expect(syncInterval, equals(syncInterval));
-        expect(syncMaxRecords, equals(mockSyncMaxRecords));
-        expect(syncPageSize, equals(mockSyncPagesize));
+                .toList());
+        expect(syncExpressions.map((expression) {
+          // Ignore generated ID
+          (expression as Map).remove("id");
+          return expression;
+        }), [expectedBlogExpression, expectedPostExpression]);
+        expect(syncInterval, mockSyncInterval);
```
Comment (HuiSF) [inline packages/amplify_datastore/test/amplify_datastore_test.dart:73]: 🙏

**Review by HuiSF — APPROVED** [review]
Overall looks great. 1 comment

File: packages/amplify_datastore/ios/Classes/SwiftAmplifyDataStorePlugin.swift (hunk 4 above)
Comment (cshfang) [reply packages/amplify_datastore/ios/Classes/SwiftAmplifyDataStorePlugin.swift:107]: Yes, excellent suggestion!

**Review by ragingsquirrel3 — COMMENTED** [review]
Thanks for adding this feature! Mostly looking great, just have a few questions (mostly noob questions) and I think I will sync offline w you regarding them.

File: packages/amplify_datastore/test/amplify_datastore_test.dart [hunk 6]
```diff
@@ -40,21 +48,31 @@ void main() {
...
   test('DataStore custom configuration should be passed via MethodChannel',
       () async {
+    var expectedBlogExpression =
+        await getJsonFromFile('sync_expressions/blog_name.json');
+    var expectedPostExpression =
+        await getJsonFromFile('sync_expressions/post_title.json');
     dataStoreChannel.setMockMethodCallHandler((MethodCall methodCall) async {
       if (methodCall.method == "configureDataStore") {
         final modelSchemas = methodCall.arguments['modelSchemas'];
+        final syncExpressions = methodCall.arguments['syncExpressions'];
```
[the remark this answers is not in the document]
Comment (ragingsquirrel3) [reply packages/amplify_datastore/test/amplify_datastore_test.dart:58]: discussed offline, maybe not necessary

**Review by ragingsquirrel3 — APPROVED** [review]

File: packages/amplify_datastore/ios/Classes/SwiftAmplifyDataStorePlugin.swift [hunk 7]
```diff
@@ -77,36 +78,63 @@ public class SwiftAmplifyDataStorePlugin: NSObject, FlutterPlugin {
...
             return //TODO
         }
-
+        
+        let syncExpressionList = args["syncExpressions"] as! [[String: Any]]
         let syncInterval = args["syncInterval"] as? Double ?? DataStoreConfiguration.defaultSyncInterval
         let syncMaxRecords = args["syncMaxRecords"] as? UInt ?? DataStoreConfiguration.defaultSyncMaxRecords
         let syncPageSize = args["syncPageSize"] as? UInt ?? DataStoreConfiguration.defaultSyncPageSize
-
+        let controller = UIApplication.shared.windows.first { $0.isKeyWindow }!.rootViewController as! FlutterViewController
```
[the remark this answers is not in the document]
Comment (cshfang) [reply packages/amplify_datastore/ios/Classes/SwiftAmplifyDataStorePlugin.swift]: Consolidated this in the next commit, thanks!

File: packages/amplify_datastore/ios/Classes/SwiftAmplifyDataStorePlugin.swift [hunk 8]
```diff
@@ -77,36 +78,63 @@ public class SwiftAmplifyDataStorePlugin: NSObject, FlutterPlugin {
...
     private func onConfigureDataStore(args: [String: Any], result: @escaping FlutterResult) {
-
+        
         guard let modelSchemaList = args["modelSchemas"] as? [[String: Any]] else {
             result(false)
             return //TODO
         }
-
+        
+        let syncExpressionList = args["syncExpressions"] as! [[String: Any]]
```
[the remark this answers is not in the document]
Comment (cshfang) [reply packages/amplify_datastore/ios/Classes/SwiftAmplifyDataStorePlugin.swift]: It should be in the map as long as we're the ones calling it but added a fallback to match Android

**Review by HuiSF — APPROVED** [review]
Thanks for adding this feature!

**Pull request closed without merging**